src/modules/renderer.ts

```diff
@@ -0,0 +1,169 @@
+    });
+    canvas.addEventListener('mousemove', e => {
+      if (this.mouseDown) {
+        mat4.rotateY(this.modelViewMatrix, this.modelViewMatrix, e.movementX * 0.01);
+        mat4.rotateX(this.modelViewMatrix, this.modelViewMatrix, e.movementY * 0.01);
+      }
+    });
+  }
+
+  // Adds an attribute if it is called for in the GLSL, and is non-empty
+  maybeAddAttr(name: string, source: number[]) {
+    if (source.length) {
+      const location = this.gl.getAttribLocation(this.program, name);
+      if (location !== -1) {
+        this.attributes[name] = {
+          location,
+          buffer: this.gl.createBuffer(),
+        };
+        this.gl.bindBuffer(this.gl.ARRAY_BUFFER, this.attributes[name].buffer);
+        this.gl.bufferData(this.gl.ARRAY_BUFFER, new Float32Array(source), this.gl.STATIC_DRAW);
+      }
+    }
+  }
+
+  // Creates a new program in the same GL context with a new model and shaders
+  // Assumes model entries length is consistent
+  compile(fragText: string, vertText: string, model: Model) {
+    // Delete the old program
+    this.gl.deleteProgram(this.program);
+
+    // Create new program
+    this.program = glHelpers.createProgram(this.gl, vertText, fragText);
+
+    this.modelLength = model.length;
+
+    // Sets attributes up if they exist in the shader code and model
+    this.maybeAddAttr('position', model.positions);
+    this.maybeAddAttr('normal', model.normals);
+
+    // Enable uniforms
+    this.uniforms = {
+      projectionMatrix: this.gl.getUniformLocation(this.program, 'projectionMatrix'),
+      modelViewMatrix: this.gl.getUniformLocation(this.program, 'modelViewMatrix'),
+      normalMatrix: this.gl.getUniformLocation(this.program, 'normalMatrix'),
+    };
+  }
+
+  // Starts a render loop
+  startRender() {
+    this.frame = requestAnimationFrame(this.render);
+  }
+
+  // Stops it
+  stopRender() {
+    cancelAnimationFrame(this.frame);
+  }
+
+  // Arrow function so that `this` works
+  render = (time: number) => {
+    // Calculate timestep, which requestAnimationFrame gives us
+    const passed = (time - this.lastTime) * 0.001;
+    this.lastTime = time;
+
+    // Resize canvas
+    glHelpers.resizeCanvas(this.gl.canvas);
+    this.gl.viewport(0, 0, this.gl.canvas.width, this.gl.canvas.height);
+
+    // Clear the canvas
+    this.gl.clear(this.gl.COLOR_BUFFER_BIT | this.gl.DEPTH_BUFFER_BIT);
+
+    // Enable stuff
+    this.gl.enable(this.gl.CULL_FACE);
+    this.gl.enable(this.gl.DEPTH_TEST);
+
+    // Use current program
+    this.gl.useProgram(this.program);
+    
+    // Iterates over attributes we previously defined, and sets them up
+    // Assumes things such as size and type which can be refined later if required
+    for (const attr of Object.values(this.attributes)) {
+      this.gl.enableVertexAttribArray(attr.location);
+      this.gl.bindBuffer(this.gl.ARRAY_BUFFER, attr.buffer);
+      this.gl.vertexAttribPointer(attr.location, 3, this.gl.FLOAT, false, 0, 0);
```

Comment:
This isn't as efficient as the old WebGL buffer implementation. 

I use a dedicated buffer for each attribute here, instead of interleaving them like the old code used to.

However, I think this version has much cleaner and more maintainable code, and considering the simplicity of our scenes, the impact of doing this instead of the fancy one buffer/interleaved version is going to be between none and negligible.